# Generate the table for displaying Pango X events

In [1]:
import numpy as np
import pandas as pd
import tszip
import sc2ts

In [2]:
ts = tszip.load("../data/sc2ts_viridian_v2.2.trees.tsz")
df_node = sc2ts.node_data(ts, inheritance_stats=False)

In [3]:
dfe = pd.read_csv("../data/pango_x_events.csv")
dfe

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,pango
0,1840333,XCT,3,R,57,{},1840333,0.0,0.000000,"{'XCT.1': 32, 'XCT': 25}",XCT
1,1480504,XAB,1,I,3,{},1251095,7.0,42.022492,"{'XQ': 58, 'BA.2': 42, 'XAM': 26, 'XAA': 22, '...",XAB
2,1117665,XG,3,I,159,{},1093814,3.0,5.781627,"{'XG': 159, 'BA.2': 11}",XG
3,1500071,XAN,0,R,26,{},1500071,0.0,0.000000,{'XAN': 26},XAN
4,1129760,XE,4,I,1256,{},1073548,2.0,9.456944,"{'XE': 1256, 'XH': 40, 'BA.2': 34}",XE
...,...,...,...,...,...,...,...,...,...,...,...
89,1782627,XBB.1.5,2,R,2,{},1782627,0.0,0.000000,{'XBB.1.5': 2},XBB
90,1796621,XBB.2.3,1,R,1,{},1796621,0.0,0.000000,{'XBB.2.3': 1},XBB
91,1793239,XBB.1.5,3,R,5,{},1793239,0.0,0.000000,{'XBB.1.5': 5},XBB
92,1804167,XBB.1,3,R,1,{},1804167,0.0,0.000000,{'XBB.1': 1},XBB


In [4]:
assert len(set(dfe["root"])) == dfe.shape[0]

In [5]:
# Convert dictionaries to Python objects.
conv1 = {}
conv2 = {}
descendants = {}
for k, row in dfe.iterrows():
    conv1[k] = eval(row["closest_recombinant_descendants"])
    conv2[k] = eval(row["non_pango_samples"])
    descendants[k] = {row["pango"]: row["pango_samples"], **conv2[k]}
dfe["closest_recombinant_descendants"] = conv1
dfe["non_pango_samples"] = conv2
dfe["descendants"] = descendants


In [6]:
dfr = pd.read_csv("../data/recombinants.csv").set_index("recombinant")
dfr

,sample_id,num_descendant_samples,num_samples,distinct_sample_pango,interval_left,interval_right,num_mutations,Viridian_amplicon_scheme,Artic_primer_version,date_added,...,parent_mrca_pango,parent_mrca_scorpio,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts
recombinant,,,,,,,,,,,,,,,,,,,,,
1438470,SRR19695004,1,1,1,519,670,2,COVID-AMPLISEQ-V1,.,2022-04-15,...,B.1.1.529,Omicron (Unassigned),1333.857513,2020-10-12,True,3,2,55,False,6
1540083,ERR9939974,1,1,1,695,958,1,COVID-ARTIC-V4.1,.,2022-06-27,...,B.1.1.529,Omicron (Unassigned),1333.857513,2020-10-12,False,4,1,12,False,9
752955,SRR20259474,1,1,1,510,1222,1,COVID-AMPLISEQ-V1,.,2021-10-21,...,B.1.617.2,Delta (B.1.617.2-like),1268.244029,2020-12-16,False,2,2,19,False,5
1639578,ERR10219711,2,1,1,695,1453,2,COVID-ARTIC-V4.1,.,2022-08-30,...,B.1.1.529,Omicron (Unassigned),1333.857513,2020-10-12,False,3,0,12,False,6
1798212,ERR11184433,1,1,1,695,1545,2,COVID-ARTIC-V4.1,.,2023-03-20,...,B.1.1.529,Omicron (Unassigned),1333.857513,2020-10-12,False,12,1,43,False,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1556595,ERR12781803,6,1,1,29509,29739,1,COVID-ARTIC-V3,.,2022-07-06,...,B.1.1.529,Omicron (Unassigned),1333.857513,2020-10-12,False,6,24,2,False,6
1750096,ERR13040343,1,1,1,29509,29739,1,COVID-ARTIC-V3,.,2023-01-20,...,B.1.1.529,Omicron (Unassigned),1333.857513,2020-10-12,False,11,35,2,False,6
1756387,ERR13041473,2,1,1,29509,29739,2,COVID-ARTIC-V3,.,2023-01-30,...,B.1.1.529,Omicron (Unassigned),1333.857513,2020-10-12,False,13,36,2,False,7


In [7]:
dfr["mutations_averted"] = dfr["k1000_muts"] - dfr["num_mutations"]

In [8]:
dfe = dfe.join(dfr, "closest_recombinant")
dfe.shape

(94, 60)

## Outlier recombinants

In [9]:
df_pe = pd.read_csv("../data/pango_events_in_arg.csv").set_index("root")

In [10]:
dfr[dfr["num_descendant_samples"] > 5000]

,sample_id,num_descendant_samples,num_samples,distinct_sample_pango,interval_left,interval_right,num_mutations,Viridian_amplicon_scheme,Artic_primer_version,date_added,...,parent_mrca_scorpio,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted
recombinant,,,,,,,,,,,,,,,,,,,,,
1673437,SRR22255464,112220,10,1,22332,22577,14,COVID-ARTIC-V4.1,.,2022-09-29,...,Omicron (BA.2-like),961.384849,2021-10-19,True,6,13,10,True,21,7
1428716,SRR21340546,244601,11,2,23041,26858,4,COVID-ARTIC-V4.1,.,2022-04-09,...,Omicron (BA.2-like),961.384849,2021-10-19,False,2,6,5,True,9,5
1213517,ERR8619597,244192,1,1,26859,27382,0,COVID-ARTIC-V4.1,4.1alt,2022-02-05,...,Omicron (Unassigned),1333.857513,2020-10-12,False,4,46,3,False,5,5


In [11]:
outliers = dfr[dfr["num_descendant_samples"] > 5000].index
for u in outliers:
    if u in df_pe.index:
        print(u, df_pe.loc[u]["pango"])
    

1673437 XBB
1428716 BA.5


In [12]:
dfe[dfe["num_descendant_samples"] > 5000]

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_scorpio,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted
27,1814205,XBL,5,I,13,{},1673437,5.0,189.868325,{'too_many_to_classify': 112021},...,Omicron (BA.2-like),961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0
46,1525011,XAV,3,S,13,{},1428716,7.0,106.000001,{'too_many_to_classify': 244541},...,Omicron (BA.2-like),961.384849,2021-10-19,False,2.0,6.0,5.0,True,9.0,5.0
53,1711075,XBZ,4,S,165,{},1428716,16.0,265.000001,{'too_many_to_classify': 244541},...,Omicron (BA.2-like),961.384849,2021-10-19,False,2.0,6.0,5.0,True,9.0,5.0
73,1841699,XDF,6,S,1,{},1673437,12.0,481.558324,{'too_many_to_classify': 112021},...,Omicron (BA.2-like),961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0
76,1814047,XDB,7,S,1,{},1673437,11.0,334.558324,{'too_many_to_classify': 112021},...,Omicron (BA.2-like),961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0
77,1814888,XDB,2,S,12,{},1673437,11.0,337.558324,{'too_many_to_classify': 112021},...,Omicron (BA.2-like),961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0
81,1673437,XBB,15,R,112005,"{'BQ.1.1.13': 2, 'JN.1.1': 11, 'JN.3': 1, 'JN....",1673437,0.0,0.000000,{'too_many_to_classify': 112021},...,Omicron (BA.2-like),961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0


So, the nodes we need to keep track of as closest recombinants that may not really be recombinants are the XBB root and BA.5 root.

It's also useful to keep track of the XQ root, as this is a complex node we need to special case.

In [13]:
df_pe[df_pe["pango"] == "XQ"]

,pango,total_samples,max_descendants,min_descendants,parent_max_descendants,parent_min_descendants,diff,diff_parent,relative_diff,relative_diff_parent
root,,,,,,,,,,
1251095,XQ,58,58,58,58,58,0,0,0.0,0.0


In [14]:
xbb_root = 1673437
ba5_root = 1428716
xq_root = 1251095

## Number of pango X lineages in the ARG



In [15]:
all_x_events = {k: int(v) for k,v in dict(dfe["root_pango"].value_counts()).items()}
all_x_events

{'XBB.1.5': 5,
 'XAC': 4,
 'XBB.1': 4,
 'XM': 3,
 'XEB': 2,
 'XAP': 2,
 'XDP': 2,
 'XDB': 2,
 'XCT': 1,
 'XAB': 1,
 'XG': 1,
 'XAN': 1,
 'XE': 1,
 'XW': 1,
 'XAZ': 1,
 'XC': 1,
 'XBV': 1,
 'XBD': 1,
 'XBS': 1,
 'XBQ': 1,
 'XF': 1,
 'XH': 1,
 'XBE': 1,
 'XBG': 1,
 'XBH': 1,
 'XAE': 1,
 'XBM': 1,
 'XCP': 1,
 'XZ': 1,
 'XBC.1.2.1': 1,
 'XBR': 1,
 'XBL': 1,
 'XR': 1,
 'XS': 1,
 'XBW': 1,
 'XAJ': 1,
 'XCF': 1,
 'XAM': 1,
 'XBK': 1,
 'XCZ': 1,
 'XL': 1,
 'XDA': 1,
 'XDD': 1,
 'XAD': 1,
 'XY': 1,
 'XAV': 1,
 'XCE': 1,
 'XP': 1,
 'XAU': 1,
 'XAG': 1,
 'XQ': 1,
 'XDK': 1,
 'XBZ': 1,
 'XB': 1,
 'XAS': 1,
 'XU': 1,
 'XV': 1,
 'XN': 1,
 'XDR': 1,
 'XCL': 1,
 'XBF': 1,
 'XCH': 1,
 'XJ': 1,
 'XDV.1': 1,
 'XDY': 1,
 'XCY': 1,
 'XA': 1,
 'XCQ': 1,
 'XDF': 1,
 'XAY.2': 1,
 'XAA': 1,
 'XCC': 1,
 'XCR': 1,
 'XAL': 1,
 'XBB': 1,
 'XBB.1.5.16': 1,
 'XBB.2.3': 1,
 'XBB.1.16.4': 1}

In [16]:
len(all_x_events)

78

In [17]:
xs = [x for x in df_node.pango.unique() if x.startswith("X") and "." not in x]
len(xs)

70

## X lineages with only sublineage samples

XDV, XAY and XBC have no samples assigned to the lineage itself in the ARG, only to its sublineages, so the origin node is assigned to a sublineage. These origin nodes are recombination nodes like any other Type I event, so we name them for the X lineage rather than dropping them with the derived lineages below.

In [18]:
sublineage_only = {"XDV.1": "XDV", "XAY.2": "XAY", "XBC.1.2.1": "XBC"}
assert not df_node["pango"].isin(sublineage_only.values()).any()
dfe["root_pango"] = dfe["root_pango"].replace(sublineage_only)
dfe[dfe["root_pango"].isin(sublineage_only.values())]

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_scorpio,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted
25,1727007,XBC,32,R,98,{},1727007,0.0,0.0,{'XBC.1.2.1': 98},...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
64,1891345,XDV,8,R,32,{},1891345,0.0,0.0,{'XDV.1': 32},...,Omicron (BA.2-like),961.384849,2021-10-19,True,9.0,27.0,43.0,True,27.0,17.0
74,1723586,XAY,17,R,371,{},1723586,0.0,0.0,"{'XAY.2.3': 127, 'XAY.2': 91, 'XAY.2.1': 81, '...",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Where are the other XDV samples?

XDV is a special case, in that the Viridian dataset does have samples assigned to XDV itself. None of them are in the ARG.

In [19]:
ds = sc2ts.Dataset("../data/viridian_mafft_2024-10-14_v1.vcz.zip", date_field="Date_tree")
df_xdv = ds.metadata.as_dataframe(["Viridian_pangolin", "Viridian_pangolin_1.29", "Date_tree"])
df_xdv = df_xdv[df_xdv["Viridian_pangolin_1.29"].str.startswith("XDV")].copy()

df_sample = df_node[df_node["is_sample"]]
df_xdv["in_arg"] = df_xdv.index.isin(df_sample["sample_id"])
df_xdv["date"] = "in ARG date range"
df_xdv.loc[df_xdv["Date_tree"] > str(df_sample["date"].max().date()), "date"] = "after ARG end date"
df_xdv.loc[df_xdv["Date_tree"] == ".", "date"] = "missing"

df_xdv.groupby(["Viridian_pangolin_1.29", "in_arg", "date"]).size()

Viridian_pangolin_1.29  in_arg  date             
XDV                     False   in ARG date range     7
                                missing               5
XDV.1                   False   in ARG date range    12
                                missing              19
                        True    in ARG date range    32
dtype: int64

The XDV assignments are from the later pangolin version; in the original Viridian assignment these samples are mostly XBB.1.19.1, one of the parent lineages.

In [20]:
pd.crosstab(df_xdv["Viridian_pangolin"], df_xdv["Viridian_pangolin_1.29"])

Viridian_pangolin_1.29,XDV,XDV.1
Viridian_pangolin,,
BA.2.10,0,3
BA.2.10.1,0,4
XBB,3,3
XBB.1.19.1,9,53


# Remove awkward events

First remove the derived lineages

In [21]:
derived_lins = [x for x in list(dfe["root_pango"]) if "." in x]
derived_lins

['XBB.1',
 'XBB.1',
 'XBB.1',
 'XBB.1.5',
 'XBB.1.5',
 'XBB.1.5.16',
 'XBB.1.5',
 'XBB.1.5',
 'XBB.2.3',
 'XBB.1.5',
 'XBB.1',
 'XBB.1.16.4']

In [22]:
pango_events_tbl = dfe[~dfe["root_pango"].isin(derived_lins)]
pango_events_tbl

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_scorpio,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted
0,1840333,XCT,3,R,57,{},1840333,0.0,0.000000,"{'XCT.1': 32, 'XCT': 25}",...,Omicron (BA.2-like),961.384849,2021-10-19,True,19.0,27.0,34.0,True,29.0,26.0
1,1480504,XAB,1,I,3,{},1251095,7.0,42.022492,"{'XQ': 58, 'BA.2': 42, 'XAM': 26, 'XAA': 22, '...",...,Omicron (Unassigned),1333.857513,2020-10-12,True,5.0,7.0,44.0,True,5.0,4.0
2,1117665,XG,3,I,159,{},1093814,3.0,5.781627,"{'XG': 159, 'BA.2': 11}",...,Omicron (Unassigned),1333.857513,2020-10-12,True,4.0,8.0,41.0,True,7.0,7.0
3,1500071,XAN,0,R,26,{},1500071,0.0,0.000000,{'XAN': 26},...,Omicron (BA.2-like),961.384849,2021-10-19,True,4.0,5.0,18.0,True,7.0,5.0
4,1129760,XE,4,I,1256,{},1073548,2.0,9.456944,"{'XE': 1256, 'XH': 40, 'BA.2': 34}",...,Omicron (Unassigned),1333.857513,2020-10-12,True,3.0,14.0,36.0,True,16.0,13.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
77,1814888,XDB,2,S,12,{},1673437,11.0,337.558324,{'too_many_to_classify': 112021},...,Omicron (BA.2-like),961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0
78,1766388,XCC,0,R,15,{},1766388,0.0,0.000000,{'XCC': 15},...,Omicron (BA.2-like),961.384849,2021-10-19,True,12.0,9.0,28.0,True,10.0,9.0
79,1833719,XCR,2,R,84,{'XBB.1.5.70': 1},1833719,0.0,0.000000,"{'XCR': 84, 'XBB.1.5.70': 1}",...,Omicron (XBB.1-like),695.299735,2022-07-12,True,8.0,18.0,11.0,True,12.0,10.0
80,1502277,XAL,2,I,14,{},1178986,4.0,70.756330,"{'XM': 31, 'XAL': 14, 'BA.2': 7}",...,Omicron (Unassigned),1333.857513,2020-10-12,True,3.0,25.0,23.0,True,9.0,8.0


## Multi-origin non-recombination events

In [23]:
def is_non_recombinant(df):
    return (
        df["closest_recombinant"].isin([-1, xbb_root, ba5_root]) & 
        (df["closest_recombinant_path_len"] > 3)
    )

pango_events_tbl[is_non_recombinant(pango_events_tbl)]["root_pango"].value_counts()

root_pango
XDB    2
XBQ    1
XBL    1
XAJ    1
XBK    1
XAV    1
XP     1
XAU    1
XBZ    1
XAS    1
XN     1
XDF    1
Name: count, dtype: int64

XBD is the only example, so we can treat specially.

In [24]:
pango_events_tbl[pango_events_tbl["root_pango"] == "XDB"]

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_scorpio,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted
76,1814047,XDB,7,S,1,{},1673437,11.0,334.558324,{'too_many_to_classify': 112021},...,Omicron (BA.2-like),961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0
77,1814888,XDB,2,S,12,{},1673437,11.0,337.558324,{'too_many_to_classify': 112021},...,Omicron (BA.2-like),961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0


In [25]:
pango_events_tbl = pango_events_tbl[
    ~
    ((pango_events_tbl["root_pango"] == "XDB") & (pango_events_tbl["pango_samples"] == 1))]
xdb_tab = pango_events_tbl[pango_events_tbl["root_pango"] == "XDB"]
assert xdb_tab.shape[0] == 1

## Multi-origin recombination events

We need to look at the multi-origin events manually, as some are really a shared event that has been labelled something else by Pangolin, and others are genuine multiple events with, in which we usually pick the majority.

In [26]:
counts = pango_events_tbl["root_pango"].value_counts() 
multiple = counts[counts > 1]
multiple

root_pango
XAC    4
XM     3
XEB    2
XAP    2
XDP    2
Name: count, dtype: int64

In [27]:
dfe_multi = dfe[dfe["root_pango"].isin(multiple.index)]
dfe_multi.sort_values("closest_recombinant")

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_scorpio,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted
36,1095085,XM,0,R,1,{'BA.1': 1},1095085,0.0,0.000000,"{'XM': 1, 'BA.1': 1}",...,Omicron (Unassigned),1333.857513,2020-10-12,True,3.0,30.0,15.0,True,30.0,30.0
67,1447906,XAP,1,I,17,{},1133051,3.0,50.446317,"{'BA.2': 218, 'XZ': 51, 'XAP': 21}",...,Omicron (Unassigned),1333.857513,2020-10-12,False,4.0,43.0,5.0,True,7.0,7.0
68,1455454,XAP,1,S,4,{},1133051,4.0,98.000000,"{'BA.2': 218, 'XZ': 51, 'XAP': 21}",...,Omicron (Unassigned),1333.857513,2020-10-12,False,4.0,43.0,5.0,True,7.0,7.0
37,1178986,XM,0,R,31,"{'BA.2': 7, 'XAL': 14}",1178986,0.0,0.000000,"{'XM': 31, 'XAL': 14, 'BA.2': 7}",...,Omicron (Unassigned),1333.857513,2020-10-12,True,3.0,25.0,23.0,True,9.0,8.0
38,1357484,XM,1,R,1,{},1357484,0.0,0.000000,{'XM': 1},...,Omicron (Unassigned),1333.857513,2020-10-12,True,3.0,29.0,21.0,True,5.0,4.0
15,1448912,XAC,2,S,1,{},1448913,2.0,80.019366,{'XAC': 19},...,Omicron (Unassigned),1333.857513,2020-10-12,True,4.0,45.0,4.0,True,7.0,4.0
16,1478564,XAC,1,S,1,{},1448913,2.0,100.019366,{'XAC': 19},...,Omicron (Unassigned),1333.857513,2020-10-12,True,4.0,45.0,4.0,True,7.0,4.0
13,1462028,XAC,1,I,8,{},1448913,2.0,30.438818,{'XAC': 19},...,Omicron (Unassigned),1333.857513,2020-10-12,True,4.0,45.0,4.0,True,7.0,4.0
14,1488108,XAC,2,I,9,{},1448913,2.0,71.979366,{'XAC': 19},...,Omicron (Unassigned),1333.857513,2020-10-12,True,4.0,45.0,4.0,True,7.0,4.0
40,1873501,XEB,2,S,5,{},1868714,2.0,53.339857,"{'XDP': 62, 'XDP.1': 24, 'XDP.3': 13, 'XEB': 5...",...,Omicron (BA.2-like),961.384849,2021-10-19,True,12.0,71.0,6.0,True,7.0,6.0


In [28]:
multi_event_names = {
    1178986: "XM/XAL", # Dominant XM event
    1133051: "XAP/XZ",
    1448913: "XAC",
    1868714: "XDP/XEB1",
    1874983: "XDP/XEB2"
}
assert set(multi_event_names.keys()) < set(dfe_multi["closest_recombinant"])

In [29]:
dfe_multi["pango_samples"].sum()

np.int64(210)

We've still only got 210 samples involved in this messiness.

The ones we want to remove are those that are in the multiple table, where the recombinant is not in one of the named events. This is just the minor XM events.

In [30]:
multi_remove = dfe_multi[~dfe_multi["closest_recombinant"].isin(multi_event_names)]
multi_remove

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_scorpio,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted
36,1095085,XM,0,R,1,{'BA.1': 1},1095085,0.0,0.0,"{'XM': 1, 'BA.1': 1}",...,Omicron (Unassigned),1333.857513,2020-10-12,True,3.0,30.0,15.0,True,30.0,30.0
38,1357484,XM,1,R,1,{},1357484,0.0,0.0,{'XM': 1},...,Omicron (Unassigned),1333.857513,2020-10-12,True,3.0,29.0,21.0,True,5.0,4.0


In [31]:
pango_events_tbl = pango_events_tbl[~pango_events_tbl.index.isin(multi_remove.index)]
pango_events_tbl

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_scorpio,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted
0,1840333,XCT,3,R,57,{},1840333,0.0,0.000000,"{'XCT.1': 32, 'XCT': 25}",...,Omicron (BA.2-like),961.384849,2021-10-19,True,19.0,27.0,34.0,True,29.0,26.0
1,1480504,XAB,1,I,3,{},1251095,7.0,42.022492,"{'XQ': 58, 'BA.2': 42, 'XAM': 26, 'XAA': 22, '...",...,Omicron (Unassigned),1333.857513,2020-10-12,True,5.0,7.0,44.0,True,5.0,4.0
2,1117665,XG,3,I,159,{},1093814,3.0,5.781627,"{'XG': 159, 'BA.2': 11}",...,Omicron (Unassigned),1333.857513,2020-10-12,True,4.0,8.0,41.0,True,7.0,7.0
3,1500071,XAN,0,R,26,{},1500071,0.0,0.000000,{'XAN': 26},...,Omicron (BA.2-like),961.384849,2021-10-19,True,4.0,5.0,18.0,True,7.0,5.0
4,1129760,XE,4,I,1256,{},1073548,2.0,9.456944,"{'XE': 1256, 'XH': 40, 'BA.2': 34}",...,Omicron (Unassigned),1333.857513,2020-10-12,True,3.0,14.0,36.0,True,16.0,13.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
77,1814888,XDB,2,S,12,{},1673437,11.0,337.558324,{'too_many_to_classify': 112021},...,Omicron (BA.2-like),961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0
78,1766388,XCC,0,R,15,{},1766388,0.0,0.000000,{'XCC': 15},...,Omicron (BA.2-like),961.384849,2021-10-19,True,12.0,9.0,28.0,True,10.0,9.0
79,1833719,XCR,2,R,84,{'XBB.1.5.70': 1},1833719,0.0,0.000000,"{'XCR': 84, 'XBB.1.5.70': 1}",...,Omicron (XBB.1-like),695.299735,2022-07-12,True,8.0,18.0,11.0,True,12.0,10.0
80,1502277,XAL,2,I,14,{},1178986,4.0,70.756330,"{'XM': 31, 'XAL': 14, 'BA.2': 7}",...,Omicron (Unassigned),1333.857513,2020-10-12,True,3.0,25.0,23.0,True,9.0,8.0


In [32]:
pango_events_tbl[pango_events_tbl["root_pango"] == "XM"]

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_scorpio,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted
37,1178986,XM,0,R,31,"{'BA.2': 7, 'XAL': 14}",1178986,0.0,0.0,"{'XM': 31, 'XAL': 14, 'BA.2': 7}",...,Omicron (Unassigned),1333.857513,2020-10-12,True,3.0,25.0,23.0,True,9.0,8.0


In [33]:
sorted(list(pango_events_tbl["root_pango"]))

['XA',
 'XAA',
 'XAB',
 'XAC',
 'XAC',
 'XAC',
 'XAC',
 'XAD',
 'XAE',
 'XAG',
 'XAJ',
 'XAL',
 'XAM',
 'XAN',
 'XAP',
 'XAP',
 'XAS',
 'XAU',
 'XAV',
 'XAY',
 'XAZ',
 'XB',
 'XBB',
 'XBC',
 'XBD',
 'XBE',
 'XBF',
 'XBG',
 'XBH',
 'XBK',
 'XBL',
 'XBM',
 'XBQ',
 'XBR',
 'XBS',
 'XBV',
 'XBW',
 'XBZ',
 'XC',
 'XCC',
 'XCE',
 'XCF',
 'XCH',
 'XCL',
 'XCP',
 'XCQ',
 'XCR',
 'XCT',
 'XCY',
 'XCZ',
 'XDA',
 'XDB',
 'XDD',
 'XDF',
 'XDK',
 'XDP',
 'XDP',
 'XDR',
 'XDV',
 'XDY',
 'XE',
 'XEB',
 'XEB',
 'XF',
 'XG',
 'XH',
 'XJ',
 'XL',
 'XM',
 'XN',
 'XP',
 'XQ',
 'XR',
 'XS',
 'XU',
 'XV',
 'XW',
 'XY',
 'XZ']

XB's origin is two edges below a multi-parent recombination node, which we name as a Type II event below.

In [34]:
pango_events_tbl.shape[0]

79

# Assigning types to each event



In [35]:
pango_events_tbl["type"] = -1
type_1_row = (
    (pango_events_tbl["root_type"] == "R") &
    (~pango_events_tbl["closest_recombinant"].isin(multi_event_names))
)
pango_events_tbl.loc[type_1_row, "type"] = 1

type_4_row = is_non_recombinant(pango_events_tbl)
pango_events_tbl.loc[type_4_row, "type"] = 4

type_3_row = pango_events_tbl["type"] == -1
pango_events_tbl.loc[type_3_row, "type"] = 3

In [36]:
type_3_recombs = set(pango_events_tbl[pango_events_tbl["type"] == 3]["closest_recombinant"])
type_3_recombs

{242304,
 463545,
 1073548,
 1093814,
 1133051,
 1178986,
 1251095,
 1448913,
 1767510,
 1868714,
 1874983}

In [37]:
assert ba5_root not in type_3_recombs
assert xbb_root not in type_3_recombs
unclassified = type_3_recombs - set(multi_event_names.keys())
assert xq_root in unclassified
unclassified.remove(xq_root)

In [38]:
pango_events_tbl[pango_events_tbl["closest_recombinant"].isin(unclassified)].sort_values("closest_recombinant")

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted,type
54,242303,XB,2,I,202,{},242304,2.0,2.000888e-10,"{'B.1': 224, 'XB': 202}",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3
7,468298,XC,1,I,5,{},463545,1.0,1.987689e+01,"{'XC': 5, 'B.1.617.2': 1}",...,1591.661216,2020-01-28,True,5.0,40.0,15.0,True,37.0,36.0,3
17,1073547,XH,3,S,40,{},1073548,1.0,8.000000e+00,"{'XE': 1256, 'XH': 40, 'BA.2': 34}",...,1333.857513,2020-10-12,True,3.0,14.0,36.0,True,16.0,13.0,3
4,1129760,XE,4,I,1256,{},1073548,2.0,9.456944e+00,"{'XE': 1256, 'XH': 40, 'BA.2': 34}",...,1333.857513,2020-10-12,True,3.0,14.0,36.0,True,16.0,13.0,3
2,1117665,XG,3,I,159,{},1093814,3.0,5.781627e+00,"{'XG': 159, 'BA.2': 11}",...,1333.857513,2020-10-12,True,4.0,8.0,41.0,True,7.0,7.0,3
32,1821991,XCF,5,I,45,{},1767510,4.0,2.317969e+01,"{'XCF': 38, 'XBB.1.5': 6, 'XCF.3': 6, 'XCF.1':...",...,1333.857513,2020-10-12,True,5.0,25.0,1.0,False,5.0,4.0,3


In [39]:
remaining_event_names = {
    463545: "XC",
    1073548: "XE/XH",
    1093814: "XG",
    1767510: "XCF",
    242304: "XB",
}
assert set(remaining_event_names.keys()) == set(unclassified)

In [40]:
type_2_event_names = multi_event_names | remaining_event_names
type_2_event_names

{1178986: 'XM/XAL',
 1133051: 'XAP/XZ',
 1448913: 'XAC',
 1868714: 'XDP/XEB1',
 1874983: 'XDP/XEB2',
 463545: 'XC',
 1073548: 'XE/XH',
 1093814: 'XG',
 1767510: 'XCF',
 242304: 'XB'}

In [41]:
all_event_names = dict(type_2_event_names)
for _, row in pango_events_tbl[pango_events_tbl["type"] == 1].iterrows():
    u = row["closest_recombinant"]
    assert u not in all_event_names
    all_event_names[u] = row["root_pango"]
all_event_names

{1178986: 'XM/XAL',
 1133051: 'XAP/XZ',
 1448913: 'XAC',
 1868714: 'XDP/XEB1',
 1874983: 'XDP/XEB2',
 463545: 'XC',
 1073548: 'XE/XH',
 1093814: 'XG',
 1767510: 'XCF',
 242304: 'XB',
 1840333: 'XCT',
 1500071: 'XAN',
 1375030: 'XW',
 1485232: 'XAZ',
 1719442: 'XBV',
 1667113: 'XBD',
 1728703: 'XBS',
 1105581: 'XF',
 1623858: 'XBE',
 1554790: 'XBG',
 1668622: 'XBH',
 1322988: 'XAE',
 1629518: 'XBM',
 1833066: 'XCP',
 1727007: 'XBC',
 1729101: 'XBR',
 1169967: 'XS',
 1788178: 'XBW',
 1834477: 'XCZ',
 1218637: 'XL',
 1863208: 'XDA',
 1872109: 'XDD',
 1311512: 'XAD',
 1413955: 'XY',
 1818894: 'XCE',
 1251095: 'XQ',
 1871605: 'XDK',
 1300637: 'XV',
 1883456: 'XDR',
 1836320: 'XCL',
 1705711: 'XBF',
 1841603: 'XCH',
 1133210: 'XJ',
 1891345: 'XDV',
 1892436: 'XDY',
 1848677: 'XCY',
 132996: 'XA',
 1825916: 'XCQ',
 1723586: 'XAY',
 1766388: 'XCC',
 1833719: 'XCR',
 1673437: 'XBB'}

In [42]:
len(all_event_names)

52

In [43]:
# One row per named recombination node, with the information needed to filter
# on downstream. Multi-parent recombinants are not in recombinants.csv, so we
# have no mutations averted for them and count their descendants here (the
# maximum over all trees, as in recombinants.csv).
df_names = pd.DataFrame({"name": all_event_names})
df_names.index.name = "recombinant"
df_names["type"] = [2 if u in type_2_event_names else 1 for u in df_names.index]
df_names["num_parents"] = [
    len(np.unique(ts.edges_parent[ts.edges_child == u])) for u in df_names.index
]
df_names["num_descendant_samples"] = dfr["num_descendant_samples"].reindex(df_names.index)
multi_parent = df_names.index[df_names["num_descendant_samples"].isna()]
assert (df_names.loc[multi_parent, "num_parents"] > 2).all()
num_descendants = np.zeros(len(multi_parent), dtype=int)
for tree in ts.trees():
    num_descendants = np.maximum(num_descendants, [tree.num_samples(u) for u in multi_parent])
df_names.loc[multi_parent, "num_descendant_samples"] = num_descendants
df_names["num_descendant_samples"] = df_names["num_descendant_samples"].astype(int)
df_names["mutations_averted"] = dfr["mutations_averted"].reindex(df_names.index).astype("Int64")
df_names = df_names.sort_index()

df_names.to_csv("../data/pango_x_event_details.csv")
df_names[df_names["num_parents"] > 2]

,name,type,num_parents,num_descendant_samples,mutations_averted
recombinant,,,,,
242304,XB,2,3,426,<NA>
1723586,XAY,1,7,371,<NA>
1727007,XBC,1,4,98,<NA>


# Formatting the table

In [44]:

def format_parent_col(df_in, parent_col, pango_col):
    parent_pango = {}
    for k, row in df_in.iterrows():
        
        focal = int(row["root"])
        parent = row[parent_col]
        if parent == -1 or np.isnan(parent):
            continue
        parent = int(parent)
        node_row = df_node.loc[parent]
        pango = node_row["pango"]
        #print(row)
        #print(node_row)
        if node_row["is_sample"]:
            is_sample = True
        else:
            children = list(ts.edges_child[ts.edges_parent == parent])
            if focal in children:
                children.remove(focal)
            df = df_node.loc[children]
            is_sample = np.any(df["is_sample"] & (df["num_mutations"] == 0))
        if is_sample:
            parent_pango[k] = "\\textbf{" + pango + "}"
        else:
            parent_pango[k] = pango
    df_in[pango_col] = parent_pango

In [45]:
def collapse_sublineages(counts):
    """
    Merge each lineage into its ancestor when that ancestor is also present,
    e.g. {XDP: 62, XDP.1: 24} -> {XDP: 86}. Pango names a direct descendant of
    A within the same alias as A.n, so this cannot merge unrelated lineages.
    This puts counts taken from closest_recombinant_descendants on the same
    footing as the pango_samples based counts used elsewhere.
    """
    collapsed = {}
    for pango, count in counts.items():
        parent = pango.rpartition(".")[0]
        while parent and parent in counts:
            pango = parent
            parent = pango.rpartition(".")[0]
        collapsed[pango] = collapsed.get(pango, 0) + count
    return dict(sorted(collapsed.items(), key=lambda item: -item[1]))

In [46]:
format_parent_col(pango_events_tbl, "parent_left", "parent_left_pango")
format_parent_col(pango_events_tbl, "parent_right", "parent_right_pango")

## Type I

In [47]:
col_name_map = {
    "root_pango": "pango",
    "mutations_averted": "averted",
    "parent_left_pango": "left p",
    "parent_right_pango": "right p",
    "interval_left": "left",
    "interval_right": "right",
    "descendants": "descendants",
    #"pango_samples": "samples", 
    #"non_pango_samples": "nested",
}
type_1_events = pango_events_tbl[pango_events_tbl["type"] == 1].copy()
type_1_events = type_1_events.sort_values(
    ["mutations_averted", "root_pango"], ascending=[False, True]
)
type_1_events[list(col_name_map.keys())].sort_values("root_pango")

,root_pango,mutations_averted,parent_left_pango,parent_right_pango,interval_left,interval_right,descendants
69,XA,19.0,\textbf{B.1.177.18},\textbf{B.1.1.7},21256.0,21765.0,{'XA': 39}
44,XAD,5.0,\textbf{BA.2},\textbf{BA.1.1},26061.0,26530.0,{'XAD': 2}
21,XAE,5.0,\textbf{BA.2},BA.1.1,24504.0,26060.0,{'XAE': 10}
3,XAN,5.0,\textbf{BA.2},\textbf{BA.5.1.23},17823.0,19862.0,{'XAN': 26}
74,XAY,NaN,NaN,NaN,NaN,NaN,{'XAY': 371}
6,XAZ,4.0,\textbf{BA.2.5},\textbf{BA.5},3359.0,9866.0,{'XAZ': 212}
81,XBB,7.0,\textbf{BA.2.10},BM.1.1.1,22332.0,22577.0,"{'XBB': 112005, 'BQ.1.1.13': 2, 'JN.1.1': 11, ..."
25,XBC,NaN,NaN,NaN,NaN,NaN,{'XBC': 98}
9,XBD,8.0,\textbf{BA.2.75.2},\textbf{BA.5.2.1},23020.0,24620.0,{'XBD': 46}
18,XBE,4.0,\textbf{BA.5.2.6},\textbf{BE.4},22600.0,27012.0,{'XBE': 74}


In [48]:
type_1_events[list(col_name_map.keys())].to_csv("../latex/pango_x_table/type1.csv")

## Type II events

In [49]:
# reindex rather than loc: multi-parent recombinants are not in recombinants.csv
type_2_events = dfr.reindex(list(type_2_event_names))
type_2_events["alias"] = type_2_event_names
# Descendants by Pango lineage of the recombination node itself. dfe records
# these for each event's closest recombinant, and every Type II event is the
# closest recombinant of at least one event.
closest_recombinant_descendants = dict(
    zip(dfe["closest_recombinant"], dfe["closest_recombinant_descendants"]))
type_2_events["descendants"] = [
    collapse_sublineages(closest_recombinant_descendants[u])
    for u in type_2_events.index]
type_2_events = type_2_events.sort_values("mutations_averted", ascending=False)
type_2_events

,sample_id,num_descendant_samples,num_samples,distinct_sample_pango,interval_left,interval_right,num_mutations,Viridian_amplicon_scheme,Artic_primer_version,date_added,...,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted,alias,descendants
recombinant,,,,,,,,,,,,,,,,,,,,,
463545,DRR321273,6.0,1.0,1.0,26768.0,27390.0,1.0,COVID-ARTIC-V3,.,2021-08-12,...,2020-01-28,True,5.0,40.0,15.0,True,37.0,36.0,XC,"{'XC': 5, 'B.1.617.2': 1}"
1073548,ERR12888175,1330.0,1.0,1.0,10448.0,11283.0,3.0,COVID-ARTIC-V3,.,2021-12-30,...,2020-10-12,True,3.0,14.0,36.0,True,16.0,13.0,XE/XH,"{'XE': 1256, 'XH': 40, 'BA.2': 34}"
1178986,SRR18814882,52.0,1.0,1.0,20056.0,21618.0,1.0,COVID-ARTIC-V4.1,.,2022-01-29,...,2020-10-12,True,3.0,25.0,23.0,True,9.0,8.0,XM/XAL,"{'XM': 31, 'XAL': 14, 'BA.2': 7}"
1133051,ERR8146303,290.0,1.0,1.0,24504.0,26060.0,0.0,COVID-ARTIC-V4.1,4.1alt,2022-01-19,...,2020-10-12,False,4.0,43.0,5.0,True,7.0,7.0,XAP/XZ,"{'BA.2': 218, 'XZ': 51, 'XAP': 21}"
1093814,ERR12888689,170.0,1.0,1.0,8394.0,9424.0,0.0,COVID-ARTIC-V3,.,2022-01-08,...,2020-10-12,True,4.0,8.0,41.0,True,7.0,7.0,XG,"{'XG': 159, 'BA.2': 11}"
1868714,SRR27366578,108.0,1.0,1.0,26834.0,26882.0,1.0,COVID-ARTIC-V5.0-5.3.2_400,.,2023-12-06,...,2021-10-19,True,12.0,71.0,6.0,True,7.0,6.0,XDP/XEB1,"{'XDP': 100, 'XEB': 5, 'JN.1.4': 3}"
1874983,SRR27492219,34.0,1.0,1.0,26834.0,27810.0,0.0,COVID-ARTIC-V5.0-5.3.2_400,.,2023-12-26,...,2021-10-19,True,11.0,67.0,6.0,True,5.0,5.0,XDP/XEB2,"{'XDP': 28, 'XEB': 4, 'JN.1.4': 2}"
1448913,SRR19495684,19.0,1.0,1.0,25811.0,26028.0,3.0,COVID-ARTIC-V4.1,.,2022-04-20,...,2020-10-12,True,4.0,45.0,4.0,True,7.0,4.0,XAC,{'XAC': 19}
1767510,SRR23710835,52.0,1.0,1.0,26276.0,27874.0,1.0,COVID-ARTIC-V4.1,.,2023-02-13,...,2020-10-12,True,5.0,25.0,1.0,False,5.0,4.0,XCF,"{'XCF': 45, 'XBB.1.5': 7}"


In [50]:
col_name_map = {
    "alias": "alias",
    "mutations_averted": "averted",
    "parent_left_pango": "left p",
    "parent_right_pango": "right p",
    "interval_left": "left",
    "interval_right": "right",
    "descendants": "descendants",
}
type_2_events[list(col_name_map.keys())]

,alias,mutations_averted,parent_left_pango,parent_right_pango,interval_left,interval_right,descendants
recombinant,,,,,,,
463545,XC,36.0,AY.29,B.1.1.7,26768.0,27390.0,"{'XC': 5, 'B.1.617.2': 1}"
1073548,XE/XH,13.0,BA.1,BA.2.9,10448.0,11283.0,"{'XE': 1256, 'XH': 40, 'BA.2': 34}"
1178986,XM/XAL,8.0,BA.1.1,BA.2,20056.0,21618.0,"{'XM': 31, 'XAL': 14, 'BA.2': 7}"
1133051,XAP/XZ,7.0,BA.2,BA.1.17.2,24504.0,26060.0,"{'BA.2': 218, 'XZ': 51, 'XAP': 21}"
1093814,XG,7.0,BA.1.1,BA.2.9,8394.0,9424.0,"{'XG': 159, 'BA.2': 11}"
1868714,XDP/XEB1,6.0,JN.1.4,FL.15,26834.0,26882.0,"{'XDP': 100, 'XEB': 5, 'JN.1.4': 3}"
1874983,XDP/XEB2,5.0,JN.1.4,XBB.1.9.1,26834.0,27810.0,"{'XDP': 28, 'XEB': 4, 'JN.1.4': 2}"
1448913,XAC,4.0,BA.2.3,BA.1.1,25811.0,26028.0,{'XAC': 19}
1767510,XCF,4.0,XBB.1.5,BA.2.10,26276.0,27874.0,"{'XCF': 45, 'XBB.1.5': 7}"


In [51]:
type_2_events[list(col_name_map.keys())].to_csv("../latex/pango_x_table/type2.csv")

## Type III events

In [52]:
type_3_events = pango_events_tbl[pango_events_tbl["type"] == 3].copy()
type_3_events["event_name"] = [all_event_names[u] for u in type_3_events["closest_recombinant"]]
type_3_events = type_3_events.sort_values(["event_name", "pango"])
type_3_events

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted,type,event_name
13,1462028,XAC,1,I,8,{},1448913,2.0,3.043882e+01,{'XAC': 19},...,2020-10-12,True,4.0,45.0,4.0,True,7.0,4.0,3,XAC
14,1488108,XAC,2,I,9,{},1448913,2.0,7.197937e+01,{'XAC': 19},...,2020-10-12,True,4.0,45.0,4.0,True,7.0,4.0,3,XAC
15,1448912,XAC,2,S,1,{},1448913,2.0,8.001937e+01,{'XAC': 19},...,2020-10-12,True,4.0,45.0,4.0,True,7.0,4.0,3,XAC
16,1478564,XAC,1,S,1,{},1448913,2.0,1.000194e+02,{'XAC': 19},...,2020-10-12,True,4.0,45.0,4.0,True,7.0,4.0,3,XAC
67,1447906,XAP,1,I,17,{},1133051,3.0,5.044632e+01,"{'BA.2': 218, 'XZ': 51, 'XAP': 21}",...,2020-10-12,False,4.0,43.0,5.0,True,7.0,7.0,3,XAP/XZ
68,1455454,XAP,1,S,4,{},1133051,4.0,9.800000e+01,"{'BA.2': 218, 'XZ': 51, 'XAP': 21}",...,2020-10-12,False,4.0,43.0,5.0,True,7.0,7.0,3,XAP/XZ
24,1380286,XZ,1,I,51,{},1133051,4.0,5.200000e+01,"{'BA.2': 218, 'XZ': 51, 'XAP': 21}",...,2020-10-12,False,4.0,43.0,5.0,True,7.0,7.0,3,XAP/XZ
54,242303,XB,2,I,202,{},242304,2.0,2.000888e-10,"{'B.1': 224, 'XB': 202}",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3,XB
7,468298,XC,1,I,5,{},463545,1.0,1.987689e+01,"{'XC': 5, 'B.1.617.2': 1}",...,2020-01-28,True,5.0,40.0,15.0,True,37.0,36.0,3,XC
32,1821991,XCF,5,I,45,{},1767510,4.0,2.317969e+01,"{'XCF': 38, 'XBB.1.5': 6, 'XCF.3': 6, 'XCF.1':...",...,2020-10-12,True,5.0,25.0,1.0,False,5.0,4.0,3,XCF


In [53]:
col_name_map = {
    "root_pango": "pango",
    "event_name": "name",
    "closest_recombinant_path_len": "path len",
    "closest_recombinant_time": "time",
    "root_mutations": "mutations",
    "descendants": "descendants",
    #"non_pango_samples": "nested",
  
    #"closest_recombinant_descendants": "by pango",
}
type_3_events[list(col_name_map.keys())]


,root_pango,event_name,closest_recombinant_path_len,closest_recombinant_time,root_mutations,descendants
13,XAC,XAC,2.0,3.043882e+01,1,{'XAC': 8}
14,XAC,XAC,2.0,7.197937e+01,2,{'XAC': 9}
15,XAC,XAC,2.0,8.001937e+01,2,{'XAC': 1}
16,XAC,XAC,2.0,1.000194e+02,1,{'XAC': 1}
67,XAP,XAP/XZ,3.0,5.044632e+01,1,{'XAP': 17}
68,XAP,XAP/XZ,4.0,9.800000e+01,1,{'XAP': 4}
24,XZ,XAP/XZ,4.0,5.200000e+01,1,{'XZ': 51}
54,XB,XB,2.0,2.000888e-10,2,{'XB': 202}
7,XC,XC,1.0,1.987689e+01,1,{'XC': 5}
32,XCF,XCF,4.0,2.317969e+01,5,{'XCF': 45}


In [54]:
type_3_events[list(col_name_map.keys())].to_csv("../latex/pango_x_table/type3.csv")

## Type IV events

In [55]:
type_4_events = pango_events_tbl[pango_events_tbl["type"] == 4].copy()
type_4_events = type_4_events.sort_values(["root_mutations", "root_pango"], ascending=True)
type_4_events

,root,root_pango,root_mutations,root_type,pango_samples,non_pango_samples,closest_recombinant,closest_recombinant_path_len,closest_recombinant_time,closest_recombinant_descendants,...,parent_mrca_time,parent_mrca_date,is_rebar_recombinant,parent_pangonet_distance,net_min_supporting_loci_lft,net_min_supporting_loci_rgt,net_min_supporting_loci_lft_rgt_ge_4,k1000_muts,mutations_averted,type
55,1518721,XAS,1,S,84,{},-1,inf,inf,{},...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4
49,1472899,XAU,1,S,9,{},-1,inf,inf,{},...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4
34,1725053,XBK,1,I,178,{'CJ.1': 1},-1,inf,inf,{},...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4
11,1733651,XBQ,1,I,36,{},-1,inf,inf,{},...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4
58,1253868,XN,1,S,141,{},-1,inf,inf,{},...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4
48,1291412,XP,1,S,48,{},-1,inf,inf,{},...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4
77,1814888,XDB,2,S,12,{},1673437,11.0,337.558324,{'too_many_to_classify': 112021},...,961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0,4
46,1525011,XAV,3,S,13,{},1428716,7.0,106.000001,{'too_many_to_classify': 244541},...,961.384849,2021-10-19,False,2.0,6.0,5.0,True,9.0,5.0,4
53,1711075,XBZ,4,S,165,{},1428716,16.0,265.000001,{'too_many_to_classify': 244541},...,961.384849,2021-10-19,False,2.0,6.0,5.0,True,9.0,5.0,4
27,1814205,XBL,5,I,13,{},1673437,5.0,189.868325,{'too_many_to_classify': 112021},...,961.384849,2021-10-19,True,6.0,13.0,10.0,True,21.0,7.0,4


In [56]:
xbb_root, ba5_root

(1673437, 1428716)

In [57]:
tree = ts.first()
root_parent = {}
root_parent_pango = {}
for k, row in type_4_events.iterrows():  
    u = tree.parent_array[row["root"]]
    root_parent[k] = u
    root_parent_pango[k] = df_node.loc[u, "pango"]

type_4_events["root_parent"] = root_parent
type_4_events["root_parent_pango"] = root_parent_pango
format_parent_col(type_4_events, "root_parent", "root_parent_pango")


In [58]:
col_name_map = {
    "root_pango": "pango",
    "root_mutations": "mutations",
    "root_parent_pango": "parent",
    #"parent_is_observed_sample": "parent sample",
    "descendants": "descendants",
    #"closest_recombinant_descendants": "by pango",
}

type_4_events[list(col_name_map.keys())]


,root_pango,root_mutations,root_parent_pango,descendants
55,XAS,1,BA.4,{'XAS': 84}
49,XAU,1,BA.2,{'XAU': 9}
34,XBK,1,CJ.1,"{'XBK': 178, 'CJ.1': 1}"
11,XBQ,1,CJ.1,{'XBQ': 36}
58,XN,1,\textbf{BA.2},{'XN': 141}
48,XP,1,\textbf{BA.1.1},{'XP': 48}
77,XDB,2,XBB.1.16.19,{'XDB': 12}
46,XAV,3,BA.5.1.24,{'XAV': 13}
53,XBZ,4,\textbf{EF.1},{'XBZ': 165}
27,XBL,5,XBB.1.5.57,{'XBL': 13}


In [59]:
type_4_events[list(col_name_map.keys())].to_csv("../latex/pango_x_table/type4.csv")